### ISSUE TITLE
Grid does not refreshes items when displayed

### ISSUE THREAD

**Issue by tuscland** [vaadin/framework#11030]
We noticed an issue with Vaadin 8.4.2 that was not present before, and is still pending with recent versions (8.4.3, 8.4.4 and 8.4.5).

A grid is displayed with a list of objects fetched from the database through a `CallbackDataProvider`. In another page, a user makes an action that changes the state of an object and clicks the back button of their browser. A fetch to the database is made again (through the underlying `DataProvider`), but the contents of the displayed items is wrong. Oddly enough, the sort order is correctly reflecting the changed object.

The expected behavior is that the grid should display the up to date contents of the objects.

Our objects implement `equals` and `hashCode` to provide equality semantics pertaining to our domain model. Removing these overloaded functions as a temporary test restores the correct behavior of the grid, but obviously breaks our domain model.

Providing an `idGetter` to the `CallbackDataProvider` does not help either, and has no effect.

We would really appreciate if someone could help understand what is going on, since 8.4.1 we are stuck and cannot follow the latest releases.

**Comment by elmot** [comment]
Hello
Do you call `DataProvider.refreshAll()` or `DataProvider.refreshItem()` after the changes?

Any reproducible code example?

**Comment by tuscland** [comment]
Hi @elmot,

Unfortunately, calling `DataProvider.refreshAll()` does not help. I am positive that the data source is queried after the change (that is, when the grid is displayed) and that up-to-date data is returned, still, the displayed data is incoherent.

I started to work on an example, but I am currently unable to reproduce the issue. Hopefully I will find what is happening...

**Comment by tuscland** [comment]
Hello @elmot,

After much search I finally got to reproduce the bug, please find attached a project to reproduce the issue: [vaadin-grid-refresh-test.zip](https://github.com/vaadin/framework/files/2198668/vaadin-grid-refresh-test.zip).

What happens is actually easier to describe:
  - a grid is set-up with `SelectionMode.SINGLE`,
  - an item is selected and a page change occurs,
  - the user modifies the object, then hits "back" to come back to the grid,
  - the grid displays the list with an item that is selected but this item has wrong data (does not reflect the changes that just occurred).

If however the grid uses a selection mode `SelectionMode.NONE` and that we use the item click handler to detect which item is selected, the correct behavior occurs: going back to the grid shows the item properly modified.

So, in a nutshell: selected items are not refreshed as they should.

Let me know if you need more information.

Best,
Camille

**Comment by mlopezFC** [comment]
Hi! .. I was able to reproduce the issue using your project:

![grid-refresh1](https://user-images.githubusercontent.com/28312758/79907870-b5943480-83f0-11ea-9167-a53d61026ee6.gif)

And then updated it to the latest version (8.10.3) and the problem seems to be gone:

![grid-refresh2](https://user-images.githubusercontent.com/28312758/79907897-c0e76000-83f0-11ea-8237-d5dc14f69d14.gif)

So I think this can be closed.

**Comment by tuscland** [comment]
Sounds good, thanks for letting me know.
